### PR TITLE
Add test verifying no invalid trace with dynamic array

### PR DESCRIPTION
**Pull request by thk123**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [n/a] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

In trying to produce a test for #4197 I've discovered the problem originally discussed in #4153 has been fixed. So I added the test and moved the invariant to a more defensive position.

### CODE REVIEW

File: jbmc/regression/jbmc/nondet-array-size/test.desc [hunk 1]
```diff
@@ -0,0 +1,16 @@
+CORE
+TestClass.class
+--function "TestClass.minimalJbmc" --unwind 5 --max-nondet-string-length 10 --java-unwind-enum-static --trace --json-ui
+^EXIT=10$
+^SIGNAL=0$
+VERIFICATION FAILED
+--
+--
+Verify that the trace produced when creating an array whose size is determined
+by a an array input, the trace is valid.
```
Comment (tautschnig) [inline jbmc/regression/jbmc/nondet-array-size/test.desc]: "a an"

**Review by tautschnig — APPROVED** [review]
I'm quite curious which change actually fixed this problem?

File: jbmc/regression/jbmc/nondet-array-size/test.desc (hunk 1 above)
Comment (tautschnig) [inline jbmc/regression/jbmc/nondet-array-size/test.desc]: "a an"

**Review by tautschnig — APPROVED** [review]
I'm quite curious which change actually fixed the problem?

File: jbmc/regression/jbmc/nondet-array-size/test.desc [hunk 2]
```diff
@@ -0,0 +1,16 @@
+CORE
+TestClass.class
+--function "TestClass.minimalJbmc" --unwind 5 --max-nondet-string-length 10 --java-unwind-enum-static --trace --json-ui
+^EXIT=10$
+^SIGNAL=0$
+VERIFICATION FAILED
+--
+--
+Verify that the trace produced when creating an array whose size is determined
```
Comment (smowton) [inline jbmc/regression/jbmc/nondet-array-size/test.desc]: "Verify that when an array's size is determined by an array input, the resulting trace is valid"?

**Review by smowton — APPROVED** [review]
Fix English, otherwise lgtm :)

**Comment by thk123** [comment]
So moving this invariant now causes some other failures so I will do the following:
 - try this branch on top of #4197 to see if that's now fixing that
 - move the invariant back and add a regex that the value is now not unknown
 - low priority: bisect between two versions to see what fixed it

**Comment by thk123** [comment]
Note failure is `cbmc-cprover-smt2-CORE` not `cbmc-CORE`

**Comment by tautschnig** [comment]
@thk123 I suppose I should also put some effort into the work in #4197.

**Comment by thk123** [comment]
@tautschnig I've not yet checked that fixes this, and so far not got a test that passes on your branch that fails on develop so I wouldn't rush. I'll let you know if the failures here are fixed by your branch.

**Comment by thk123** [comment]
@tautschnig Still get this invariant fail on top of your branch, specifically on cbmc-cprover-smt2:

- json1/test.desc
- trace_options_json_extended/extended.desc
- trace_options_json_extended/non-extended.desc 

If you want to investigate them further. For now I will put the invariant back and see if I can make the test a little more precise.

File: src/goto-programs/json_goto_trace.cpp [hunk 3]
```diff
@@ -118,6 +118,10 @@ void convert_decl(
   const symbolt *symbol;
   irep_idt base_name, display_name;
 
+  DATA_INVARIANT(
```
Comment (peterschrammel) [inline src/goto-programs/json_goto_trace.cpp:121]: Compare with the corresponding code in goto_trace.cpp: https://github.com/diffblue/cbmc/blob/557be1d344d01cb2826deaff12e3a65222f93b03/src/goto-programs/goto_trace.cpp#L271

**Comment by tautschnig** [comment]
> @tautschnig Still get this invariant fail on top of your branch, specifically on cbmc-cprover-smt2: [...]

It seems the SMT2 solver simply is broken when it comes to unbounded/infinity-sized arrays (@kroening might have some input): The solver returns as model
```
((|__CPROVER_threads_exited#1| |__CPROVER_threads_exited#1|))
```
which is not usable as it is self-referential.

**Comment by tautschnig** [comment]
@thk123 I took the liberty to rebase your branch, include tags for the broken SMT test, and apply @smowton text suggestion. I'll leave it to you to merge this as I'm not sure whether moving the invariant would break TG in some way.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 08c7c9dfe4eaff8b40dbf386cc92059f5295cf82).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106404959

**Comment by thk123** [comment]
This passes tests locally for me, I believe I only ever saw problems on JBMC

**Pull request merged**